l1-contracts/contracts/state-transition/StateTransitionManager.sol

```diff
@@ -0,0 +1,240 @@
+    constructor(address _bridgehub) reentrancyGuardInitializer {
+        bridgehub = _bridgehub;
+    }
+
+    modifier onlyBridgehub() {
+        require(msg.sender == bridgehub, "StateTransition: only bridgehub");
+        _;
+    }
+
+    modifier onlyChain(uint256 _chainId) {
+        require(stateTransition[_chainId] == msg.sender, "StateTransition: only chain");
+        _;
+    }
+
+    modifier onlyChainGovernor(uint256 _chainId) {
+        require(
+            IZkSyncStateTransition(stateTransition[_chainId]).getGovernor() == msg.sender,
+            "StateTransition: only chain governor"
+        );
+        _;
+    }
+
+    /// @dev initialize
+    function initialize(
+        StateTransitionManagerInitializeData calldata _initializeData
+    ) external reentrancyGuardInitializer {
+        require(_initializeData.governor != address(0), "StateTransition: governor zero");
+        _transferOwnership(_initializeData.governor);
+
+        genesisUpgrade = _initializeData.genesisUpgrade;
+        protocolVersion = _initializeData.protocolVersion;
+
+        // We need to initialize the state hash because it is used in the commitment of the next batch
+        IExecutor.StoredBatchInfo memory batchZero = IExecutor.StoredBatchInfo(
+            0,
+            _initializeData.genesisBatchHash,
+            _initializeData.genesisIndexRepeatedStorageChanges,
+            0,
+            EMPTY_STRING_KECCAK,
+            DEFAULT_L2_LOGS_TREE_ROOT_HASH,
+            0,
+            _initializeData.genesisBatchCommitment
+        );
+        storedBatchZero = keccak256(abi.encode(batchZero));
+
+        initialCutHash = keccak256(abi.encode(_initializeData.diamondCut));
+
+        // While this does not provide a protection in the production, it is needed for local testing
+        // Length of the L2Log encoding should not be equal to the length of other L2Logs' tree nodes preimages
+        assert(L2_TO_L1_LOG_SERIALIZE_SIZE != 2 * 32);
+    }
+
+    /// @dev set initial cutHash
+    function setInitialCutHash(Diamond.DiamondCutData calldata _diamondCut) external onlyOwner {
+        initialCutHash = keccak256(abi.encode(_diamondCut));
+    }
+
+    /// @dev set New Version with upgrade from old version
+    function setNewVersionUpgrade(
+        Diamond.DiamondCutData calldata _cutData,
+        uint256 _oldProtocolVersion,
+        uint256 _newProtocolVersion
+    ) external onlyOwner {
+        upgradeCutHash[_oldProtocolVersion] = keccak256(abi.encode(_cutData));
+        protocolVersion = _newProtocolVersion;
+    }
+
+    /// @dev set upgrade for some protocolVersion
+    function setUpgradeDiamondCut(
+        Diamond.DiamondCutData calldata _cutData,
+        uint256 _oldProtocolVersion
+    ) external onlyOwner {
+        upgradeCutHash[_oldProtocolVersion] = keccak256(abi.encode(_cutData));
+    }
+
+    /// upgrade a specific chain
+    function upgradeChainFromVersion(
```

Comment:
The governor permissions are not clear with this flow. Governor of the State Transition chain calls `StateTransitionManager` that calls `StateTransitionChain`. So by looking into `StateTransitionChain` it seems that Governor has no power, but in fact there are other contracts that checks `StateTransitionChain.governor()` to be `msg.sender`.

Current flow:
`Governor -> StateTransitionManager -> StateTransitionChain.`

Flow I would rather want to see:
`Governor -> StateTransitionChain -> StateTransitionManager`

Reply:
The upgrade is set in STM and not STC, so the STC will have to call STM. 
So what I can do is Governor ->STC ->STM ->STC instead, if you prefer? (the current setup removes the first STC element)

Reply:
Yeah, want to make access control a bit clearer. 

`Governor ->STC ->STM ->STC` works, but I think you still can do `Governor ->STC ->STM`, `STM` then need to have a getter that return true in case upgrade is approved and false otherwise. WDYT?

Reply:
Ok, I moved upgradeChainFromVersion to STC, still kept executeUpgrade if we need it in the future